cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java

```diff
@@ -0,0 +1,219 @@
+    try {
+      if (namespaceSourceControlMetadataTable == null) {
+        namespaceSourceControlMetadataTable = context.getTable(
+            StoreDefinition.NamespaceSourceControlMetadataStore.NAMESPACE_SOURCE_CONTROL_METADATA);
+      }
+    } catch (TableNotFoundException e) {
+      throw new TableNotFoundException(
+          StoreDefinition.NamespaceSourceControlMetadataStore.NAMESPACE_SOURCE_CONTROL_METADATA);
+    }
+    return namespaceSourceControlMetadataTable;
+  }
+
+  /**
+   * Retrieves the source control metadata for the specified application ID in the namespace from
+   * {@code NamespaceSourceControlMetadata} table.
+   *
+   * @param appRef {@link ApplicationReference} for which the source control metadata is being
+   *               retrieved.
+   * @return The {@link SourceControlMeta} associated with the application ID, or {@code null} if no
+   * metadata is found.
+   * @throws IOException If it fails to read the metadata.
+   */
+  @Nullable
+  public SourceControlMeta get(ApplicationReference appRef) throws IOException {
+    List<Field<?>> primaryKey = getPrimaryKey(appRef);
+    StructuredTable table = getNamespaceSourceControlMetadataTable();
+    Optional<StructuredRow> row = table.read(primaryKey);
+
+    return row.map(nonNullRow -> {
+      String specificationHash = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD);
+      String commitId = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD);
+      Long lastSynced = nonNullRow.getLong(
+          StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD);
+      lastSynced = lastSynced == 0L ? null : lastSynced;
+       Boolean isSynced = nonNullRow.getBoolean(
+          StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD);
+      if (specificationHash == null && commitId == null && lastSynced == null) {
+        return null;
+      }
+      return new SourceControlMeta(specificationHash, commitId, Instant.ofEpochMilli(lastSynced),
+          isSynced);
+    }).orElse(null);
+  }
+
+  /**
+   * Sets the source control metadata for the specified application ID in the namespace.  Source
+   * control metadata will be null when the application is deployed in the namespace. It will be
+   * non-null when the application is pulled from the remote repository and deployed in the
+   * namespace.
+   *
+   * @param appRef            {@link ApplicationReference} for which the source control metadata is
+   *                          being set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void write(ApplicationReference appRef,
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
```

Comment:
Add `IllegalArgumentException` to the throws list

Reply:
Done

Reply:
I dont see the change